### PR TITLE
Make all DOTNET_ envvars, build envvars only

### PR DESCRIPTION
**Pull request by tmds**
Environment variables used by the assemble and run script, need to
be provided in the OpenShift build environment and deployment
environment.

We only uses the DOTNET_ environment variables in the
assemble script (build environment). The output is published
at /opt/app-root/publish including an s2i_run script to start
the .NET Core application.

### CODE REVIEW

**Comment by tmds** [comment]
@jerboaa @omajid ptal

File: 1.1/s2i/bin/run [hunk 1]
```diff
@@ -3,21 +3,14 @@ source /opt/app-root/etc/generate_container_user
...
-DOTNET_CONFIGURATION="${DOTNET_CONFIGURATION:-Release}"
-DOTNET_STARTUP_PROJECT="${DOTNET_STARTUP_PROJECT:-.}"
-
-# Private environment
-DOTNET_FRAMEWORK="netcoreapp1.1"
-APP_DLL_NAME="$(basename "$(realpath "${DOTNET_STARTUP_PROJECT}")").dll"
-APP_DLL_DIR="${DOTNET_STARTUP_PROJECT}/bin/${DOTNET_CONFIGURATION}/${DOTNET_FRAMEWORK}/publish"
+cd $DOTNET_PUBLISH_PATH
 
 ARG_FILE="arguments.txt"
```
Comment (tmds) [inline 1.1/s2i/bin/run]: @jerboaa do you know where 'arguments.txt' comes from? This is not a .NET pattern.
It isn't in the documentation. May it be removed?

File: 1.1/Dockerfile.rhel7 [hunk 2]
```diff
@@ -41,8 +43,8 @@ RUN yum install -y yum-utils && \
     yum-config-manager --enable rhel-7-server-ose-3.0-rpms && \
     yum install -y --setopt=tsflags=nodocs rh-dotnetcore11 nss_wrapper tar rh-nodejs4-npm && \
     yum clean all && \
-    mkdir -p /opt/app-root/src && \
-    useradd -u 1001 -r -g 0 -d /opt/app-root/src -s /sbin/nologin \
+    mkdir -p /opt/app-root/src /opt/app-root/publish && \
+    useradd -u 1001 -r -g 0 -d /opt/app-root -s /sbin/nologin \
       -c "Default Application User" default && \
```
Comment (tmds) [inline 1.1/Dockerfile.rhel7]: small fix: changed the -d argument to match with HOME

File: 1.1/test/dotnetbot/project.json [hunk 3]
```diff
@@ -13,5 +13,10 @@
     "netcoreapp1.1": {
       "imports": "dnxcore50"
     }
+  },
+  "publishOptions": {
+    "include": [
+      "arguments.txt"
+    ]
```
Comment (tmds) [inline 1.1/test/dotnetbot/project.json]: arguments.txt now needs to be a published file of the DOTNET_STARTUP_PROJECT
before it was a file in context-dir
To keep the old semantics, we'd need to treat it specially in the assemble script.

File: 1.1/Dockerfile.rhel7 (hunk 2 above)
Comment (jerboaa) [inline 1.1/Dockerfile.rhel7]: I'd prefer if we kept the user's home directory to be in the src folder. The reason for this is that when you get a terminal in the app container, the sources of the app will be in the current working directory. If we really want to change this, please do so in a separate PR, not lumping it in here. Thanks.

File: 1.1/s2i/bin/run (hunk 1 above)
Comment (jerboaa) [inline 1.1/s2i/bin/run]: It's just for convenience if there are many arguments to pass to an app. Pattern taken from some other container, that escapes me. It's been there from the start. If we are going to remove it we should perhaps give users some migration path.

**Comment by jerboaa** [comment]
Can you elaborate the motivation about these changes? Is it to give users a way to debug problems easier?

**Comment by tmds** [comment]
> Can you elaborate the motivation about these changes? Is it to give users a way to debug problems easier?

This makes it impossible to break the deployment by specifying DOTNET_* envvars.
Since the publish folder is outside of /src, the change makes it possible to remove the src folder.
It is also a step towards a runtime image.

File: 1.1/s2i/bin/run (hunk 1 above)
Comment (tmds) [inline 1.1/s2i/bin/run]: .NET has support for reading configuration in 'Microsoft.Extensions.Configuration'. There are providers that read Xml, Ini and Json configuration files. These are good alternatives for arguments.txt.
Remove in separate PR?

File: 1.1/Dockerfile.rhel7 (hunk 2 above)
Comment (tmds) [inline 1.1/Dockerfile.rhel7]: As it is intentional, I will revert the change.

File: 1.1/test/dotnetbot/project.json (hunk 3 above)
Comment (tmds) [inline 1.1/test/dotnetbot/project.json]: This change can be reverted when we drop support for arguments.txt.

File: 1.1/test/qotd/project.json [hunk 4]
```diff
@@ -16,7 +16,8 @@
   },
   "publishOptions": {
     "include": [
-      "quotes.txt"
+      "quotes.txt",
+      "arguments.txt"
     ]
```
Comment (tmds) [inline 1.1/test/qotd/project.json]: This change can be reverted when we drop support for arguments.txt.

File: 1.1/test/dotnetbot/project.json (hunk 3 above)
Comment (tmds) [inline 1.1/test/dotnetbot/project.json]: This change can be reverted when we drop support for arguments.txt.

File: 1.1/s2i/bin/run (hunk 1 above)
Comment (tmds) [inline 1.1/s2i/bin/run]: > Remove in separate PR?

Thinking about it -> separate PR is preferable since it will also require removal of files and changing tests.

Comment (jerboaa) [inline 1.1/s2i/bin/run]: +1

File: 1.1/Dockerfile.rhel7 (hunk 2 above)
Comment (jerboaa) [inline 1.1/Dockerfile.rhel7]: I don't see this reverted, yet :(

**Comment by jerboaa** [comment]
> This makes it impossible to break the deployment by specifying DOTNET_* envvars.
Since the publish folder is outside of /src, the change makes it possible to remove the src folder.
It is also a step towards a runtime image.

OK, this makes sense. I had to read the original commit message twice to fully grok this, though. Basically the environment variables will still be there, but they won't be used as the `run` script delegates to `s2i_run` which does not need/use them. Can I ask you to give such rationale right when posting the PR?

**Comment by tmds** [comment]
> Can I ask you to give such rationale right when posting the PR?

That was my intent. I agree, the commit message is hard to digest.

**Comment by jerboaa** [comment]
Thanks, merged.
